### PR TITLE
show politeia proposals 

### PR DESCRIPTION
**Pull request by dreacot**
Requires https://github.com/planetdecred/dcrlibwallet/pull/161

This PR allows users to;

1. View Decred politeia proposals
2. View politeia proposals details
3. View vote status of politeia proposals
4. Receive notifications for proposals on (new proposal, vote started, vote finished)

**How the notifications work**

When the client launches the app for the first time, a sync is initiated and all proposals are fetched, now when next the client launches the app the sync initiates again, and if there are new proposals or changes to the proposals fetched initially, our callback listeners are triggered

We have listeners for `onNewProposal, onVoteStarted, and onVoteEnded` . When any of the listeners are triggered, it sends the appropriate notification to the client

When the client taps on a particular notification, it takes them to the proposal details page for that particular proposal

**Screenshots**
| || |
|-|-|-|

| ||
|-|-|

### CODE REVIEW

File: app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            finish()
+        }
+    }
+
+    private fun loadProposalDetails() = GlobalScope.launch(Dispatchers.Default) {
+        val proposalResult = multiWallet!!.politeia!!.getProposalByID(proposal!!.id)
+        val proposalObjectJson = JSONObject(proposalResult).getJSONObject("result")
+        val proposalResultString: String = proposalObjectJson.toString()
+
+        val voteObjectJson = proposalObjectJson.getJSONObject("votesummary")
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt:90]: Put `votesummary` in Constants class and every other constant.

File: app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+        runOnUiThread {
+            val proposalItem = gson.fromJson(proposalResultString, Proposal::class.java)
+            val voteSummaryItem = gson.fromJson(voteSummaryString, Proposal.VoteSummary::class.java)
+
+            proposalTitle.text = proposalItem.name
+            author.text = proposal!!.username
+            timestamp.text = Utils.calculateTime(System.currentTimeMillis() / 1000 - proposal!!.timestamp, this@ProposalDetailsActivity)
+            comments.text = String.format(Locale.getDefault(), "%d Comments", proposal!!.getNumcomments())
+            version.text = String.format(Locale.getDefault(), "version %s", proposal!!.version)
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt:101]: Use string resources.

File: app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    private fun loadProposalDetails() = GlobalScope.launch(Dispatchers.Default) {
+        val proposalResult = multiWallet!!.politeia!!.getProposalByID(proposal!!.id)
+        val proposalObjectJson = JSONObject(proposalResult).getJSONObject("result")
+        val proposalResultString: String = proposalObjectJson.toString()
+
+        val voteObjectJson = proposalObjectJson.getJSONObject("votesummary")
+        val voteSummaryString: String = voteObjectJson.toString()
+
+        runOnUiThread {
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt:93]: Most of the work done in this UI thread can be done in the background and the result can be posted to the UI.

File: app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+            // Get and read file from array.
+            if (proposal!!.files != null && proposal!!.files!!.isNotEmpty()) {
+                progressBar.visibility = View.VISIBLE
+
+                var payload: String?
+                val description = StringBuilder()
+                var i = 0
+                while (proposalItem.files != null && i < proposalItem.files!!.size) {
+                    if (proposalItem.files!![i].name == "index.md") {
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt:111]: Use constants.

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.collections.ArrayList
+
+class PoliteiaActivity : BaseActivity(), SwipeRefreshLayout.OnRefreshListener, OnItemSelectedListener, ViewTreeObserver.OnScrollChangedListener {
+    private var proposals = ArrayList<Proposal>()
+    private var recyclerView: RecyclerView? = null
+    private var proposalAdapter: ProposalAdapter? = null
+    private var swipeRefreshLayout: SwipeRefreshLayout? = null
+    private var layoutManager: LinearLayoutManager? = null
+    private lateinit var emptyList: TextView
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:39]: Remove variable declarations for views and access them directing using their ids.

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+            override fun onNothingSelected(parent: AdapterView<*>?) {}
+        }
+
+        go_back.setOnClickListener {
+            finish()
+        }
+        refreshAvailableProposalCategories()
+    }
+
+    override fun onScrollChanged() {
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:122]: In this function, enable toolbar shadow if `scrollY > 0` and disable if 0.

File: app/src/main/java/com/dcrandroid/util/NetworkTask.kt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+class NetworkTask(config: HashMap<String, String>, private var delegate: AsyncResponse) : AsyncTask<String, Int, String>() {
+    private var config: HashMap<*, *> = config
+    override fun doInBackground(vararg strings: String): String? {
+        try {
+            val url = URL(if (config["url"] == null) "" else config["url"].toString())
+            val connection = url.openConnection() as HttpURLConnection
+            connection.requestMethod = if (config["method"] == null) "GET" else config["method"].toString().toUpperCase()
+            connection.doInput = true
+            connection.readTimeout = if (config["readtimeout"] == null) 7000 else Integer.valueOf(config["readtimeout"].toString())
+            connection.connectTimeout = if (config["connecttimeout"] == null) 7000 else Integer.valueOf(config["connecttimeout"].toString())
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/util/NetworkTask.kt:22]: These strings should be placed in the Constants class.

File: app/src/main/java/com/dcrandroid/util/Utils.kt [hunk 8]
```diff
@@ -217,6 +223,77 @@ object Utils {
...
+            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+            notificationManager.createNotificationChannel(channel)
+        }
+    }
+
+    fun sendProposalNotification(context: Context, manager: NotificationManager, proposalID: Long, title: String,
+                                 token: String) {
+
+        val gson = GsonBuilder().registerTypeHierarchyAdapter(ArrayList::class.java, Deserializer.ProposalDeserializer()).create()
+        val proposalResult = multiWallet!!.politeia!!.getProposalByID(proposalID)
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/util/Utils.kt:243]:
```suggestion
        val proposalResult = WalletData.multiWallet!!.politeia!!.getProposalByID(proposalID)
```

File: app/src/main/res/values-fr/strings.xml [hunk 9]
```diff
@@ -536,5 +536,7 @@
     <string name="disconnect_delete_wallet">Déconnecter/annuler la synchronisation avant de supprimer un porte-monnaie</string>
     <string name="seed_backup_success_msg">Soyez certain de conserver votre phrase de récupération dans un lieu sûr.</string>
     <string name="seed_backup_verified">Votre phrase de récupération est vérifiée</string>
+    <string name="politeia">Politeia</string>
+    <string name="proposal_details">Politeia Details</string>
```
Comment (C-ollins) [inline app/src/main/res/values-fr/strings.xml:541]: Undo this change and also changes made to every other non-English string file.

**Review by C-ollins — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/dcrandroid/HomeActivity.kt [hunk 10]
```diff
@@ -130,10 +133,24 @@ class HomeActivity : BaseActivity(), SyncProgressListener, TxAndBlockNotificatio
             currentBottomSheet = sendPageSheet
         }
 
+        syncProposals()
```
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/HomeActivity.kt:136]: isn't that a bit restricting? to some extent the app should still be usable without the need for the wallet to be online, just like checking transaction history for example

File: app/src/main/java/com/dcrandroid/HomeActivity.kt [hunk 11]
```diff
@@ -130,10 +133,24 @@ class HomeActivity : BaseActivity(), SyncProgressListener, TxAndBlockNotificatio
...
     }
 
+    private fun syncProposals() {
+        Thread(Runnable {
+            try {
+                multiWallet!!.politeia.sync(this@HomeActivity)
+                println("Logged IN Broadcast sent")
+            } catch (e: java.lang.Exception) {
+                e.printStackTrace()
+                runOnUiThread { Toast.makeText(this@HomeActivity, "PoliteiaSyncError " + e.localizedMessage, Toast.LENGTH_SHORT).show() }
```
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/HomeActivity.kt:149]: if this error hit on first run, it would tell the user no available proposal on the proposal page

should we add a retry button in settings?

File: app/src/main/java/com/dcrandroid/util/Utils.kt (hunk 8 above)
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/util/Utils.kt:243]: adding `WalletData` comes up as redundant

File: app/src/main/java/com/dcrandroid/HomeActivity.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (C-ollins) [reply app/src/main/java/com/dcrandroid/HomeActivity.kt:136]: Transaction history is offline, you can load the saved proposals if the wallet is offline.

File: app/src/main/java/com/dcrandroid/HomeActivity.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (C-ollins) [reply app/src/main/java/com/dcrandroid/HomeActivity.kt:149]: A retry button can be added and `Toast` should not be used anywhere in the app. Use the app's default `SnackBar` if you really need to display a short message for the user.

File: app/src/main/java/com/dcrandroid/HomeActivity.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/HomeActivity.kt:136]: okay noted

File: app/src/main/java/com/dcrandroid/HomeActivity.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/HomeActivity.kt:149]: okay noted

File: app/src/main/java/com/dcrandroid/util/Utils.kt (hunk 8 above)
Comment (C-ollins) [reply app/src/main/java/com/dcrandroid/util/Utils.kt:243]: It's redundant because you added `multiWallet` as a static import which is what I'm trying to avoid. Add `WalletData` and remove the `multiWallet` import.

Comment (dreacot) [reply app/src/main/java/com/dcrandroid/util/Utils.kt:243]: okay, i understand now, thanks

File: app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt (hunk 3 above)
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/activities/ProposalDetailsActivity.kt]: what do you mean done in the background?

File: app/src/main/java/com/dcrandroid/fragments/more/PoliteiaFragment.kt [hunk 12]
```diff
@@ -0,0 +1,864 @@
...
+        }
+
+        loading.set(true)
+        val limit = 10
+        val offset = when {
+            loadMore -> proposals.size
+            else -> 0
+        }
+
+        val jsonResult = multiWallet.politeia.getProposals(Constants.PROPOSALS_ABANDONED, offset, limit, newestProposalsFirst)
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/fragments/more/PoliteiaFragment.kt:641]: All the `load-` functions on this page are redundant and should be merged into one. This particular line is the only thing different in each function and you can replace it with a parameter.

File: app/src/main/java/com/dcrandroid/fragments/more/PoliteiaFragment.kt [hunk 13]
```diff
@@ -0,0 +1,864 @@
...
+    }
+
+    private fun refreshAvailableProposalCategories() = GlobalScope.launch(Dispatchers.Default) {
+        availableProposalTypes.clear()
+
+        val preCount = multiWallet.politeia.count(Constants.PROPOSALS_PRE)
+        val activeCount = multiWallet.politeia.count(Constants.PROPOSALS_ACTIVE)
+        val approvedCount = multiWallet.politeia.count(Constants.PROPOSALS_APPROVED)
+        val rejectedCount = multiWallet.politeia.count(Constants.PROPOSALS_REJECTED)
+        val abandonedCount = multiWallet.politeia.count(Constants.PROPOSALS_ABANDONED)
```
Comment (C-ollins) [inline app/src/main/java/com/dcrandroid/fragments/more/PoliteiaFragment.kt:718]: Use dcrlibwallet constants
```suggestion
        val preCount = multiWallet.politeia.count(Dcrlibwallet.ProposalCategoryPre)
        val activeCount = multiWallet.politeia.count(Dcrlibwallet.ProposalCategoryActive)
        val approvedCount = multiWallet.politeia.count(Dcrlibwallet.ProposalCategoryApproved)
        val rejectedCount = multiWallet.politeia.count(Dcrlibwallet.ProposalCategoryRejected)
        val abandonedCount = multiWallet.politeia.count(Dcrlibwallet.ProposalCategoryAbandoned)
```

**Review by C-ollins — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+        switchFragment()
+    }
+
+    private fun switchFragment() {
+        currentFragment = PoliteiaFragment()
+        supportFragmentManager
+                .beginTransaction()
+                .replace(R.id.container, currentFragment)
+                .commit()
```
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:32]: the code from the activity was moved to a fragment, so the sync function can be accessed from the base fragment

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt [hunk 15]
```diff
@@ -73,27 +91,17 @@ class PoliteiaActivity : BaseActivity(), ProposalNotificationListener,
...
-
-                loadProposals(false, proposalCategory)
+                loadProposals(false)
             }
 
             override fun onNothingSelected(parent: AdapterView<*>?) {}
         }
 
         go_back.setOnClickListener {
+            multiWallet!!.politeia.clearSavedProposals()
```
[the remark this answers is not in the document]
Comment (C-ollins) [reply app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:104]: That was for debugging, will remove it.

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt [hunk 16]
```diff
@@ -102,19 +110,55 @@ class PoliteiaActivity : BaseActivity(), ProposalNotificationListener,
...
+            setSyncButtonState()
+        }
+    }
+
+    private fun setSyncButtonState() = GlobalScope.launch(Dispatchers.Main){
+        if(multiWallet!!.politeia.isConnected){
+            tv_sync_label.setText(R.string.stop_sync)
+
+            if(multiWallet!!.politeia.synced()) {
+                sync_icon.animation = null
```
[the remark this answers is not in the document]
Comment (C-ollins) [reply app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:155]: Sync doesn't 'complete', what completed is the fetching of all unfetched proposals and dcrlibwallet will continue to check every 30 minutes for proposal updates.

This is why I don't really support that 'Sync' phrase in this part of the app because it's going to be confused with normal syncing.

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:104]: alright

File: app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (dreacot) [reply app/src/main/java/com/dcrandroid/activities/more/PoliteiaActivity.kt:155]: oh i see, yes it's a bit confusing, we might have to find a suitable synonym

**Comment by dreacot** [comment]
Reviewed and everything looks good

**Review by C-ollins — APPROVED** [review]
